File: 백준/Silver/4779. 칸토어 집합/칸토어 집합.py
def cantorian(line: str) -> str:
    if len(line) == 1:
        return line
    
    s = len(line) // 3
    prefix = line[:s]
    postfix = line[-s:]
    
    return cantorian(prefix) + " " * s + cantorian(postfix)

while True:
    try:
        n = int(input())
        line = "-" * (3 ** n)
        print(cantorian(line))
    except:
        break

File: 백준/Silver/4779. 칸토어 집합/칸토어 집합.rs
use std::io::stdin;

fn cantorian(line: String) -> String {
    if line.len() == 1 {
        return line;
    }

    let s: usize = line.len() / 3;
    let prefix = line.chars().take(s).collect::<String>();
    let suffix = line.chars().skip(2 * s).collect::<String>();
    let space = " ".repeat(s);

    return format!("{}{}{}", cantorian(prefix), space, cantorian(suffix));
}

fn main() {
    loop {
        let mut buffer = String::new();
        stdin().read_line(&mut buffer).unwrap();
        let ret = buffer.trim().parse::<usize>();
        if ret.is_err() {
            break;
        }
        buffer.clear();

        let n = ret.unwrap();
        let line = "-".repeat(3_usize.pow(n as u32));

        println!("{}", cantorian(line));
    }
}
